content/blog/integrating-python-3-with-fips-enabled-openssl-3-1.md

```diff
@@ -0,0 +1,150 @@
+
+**Step 1:** Download OpenSSL source and FIPS source from `http://www.openssl.org` and the Python source from `http://www.python.org`.
+
+**Step 2:** Install NASM and Perl Software and add to PATH
+
+**Step 3:** Install these 2 Perl modules using the commands below.
+
+> `cpan -i Text::Template`\
+> `cpan -i Test::More`
+
+**Step 4:** Create a directory structure like what's shown below.
+
+> `C:\SSLout\SSL`\
+> `C:\SSLout\DLL\x64\Release`\
+> `C:\SSLout\Lib`
+
+**Step 5:** Unzip the `openssl-3.1.0.tar.gz` file into `C:\work\openssl-3.1.0`
+
+**Step 6:** Build the OpenSSL module using `VC++ 2015 x64 the Native Tools` Command prompt.
+
+> `cd openssl-3.1.0`\
+> `Perl Configure VC-WIN64A --prefix=C:\SSLout\DLL\x64\Release --openssldir=c:\SSLout\SSL enable-fips`\
+> `nmake`\
+> `nmake install_sw`\
+> `nmake install_ssldirs`\
+> `nmake install_docs`\
+> `nmake install_fips`
+
+The OpenSSL 3.1 binaries are generated in the `C:\SSLout\DLL\x64\Release\bin`.  
+
+This is shown as shown below:
+
+![](/img/openssl_directory_structure.jpg)
+
+**Step 7:** Unzip Python in Windows to `C:\work\Python311` and go into `C:\work\Python-3.11.2\PCBuild`.  
+
+Run the `get_externals.bat` file.  This will fetch the dependencies from the internet into the `externals` directory. Now create the `openssl-bin-3.1.0` directory under the `externals` directory as shown below.
+
+![](/img/openssl_directory_structure2.jpg)
+
+**Step 8:** Under the `externals` directory, create an `amd64` depending on the CPU architecture and copy the files to it as shown below:
+
+> copy `C:\SSLout\DLL\x64\Release\lib\libcrypto.lib` and `libssl.lib` to `openssl-bin-3.1.0\amd64`\
+> copy `C:\SSLout\DLL\x64\Release\bin\*.*` to `openssl-bin-3.1.0\amd64`\
+> copy `C:\SSLout\DLL\x64\Release\include` directory to `openssl-bin-3.1.0\amd64`\
+> copy `C:\SSLout\DLL\x64\Release\include\openssl\applink.c` to `openssl-bin-3.1.0\amd64\include`\
+> copy `C:\SSLout\DLL\x64\Release\lib\ossl-modules\*.*` to `openssl-bin-3.1.0\amd64`
+
+**Step 9:** Modify `PCbuild/openssl.props` as shown below:
+
+![](/img/openssl_settings.jpg)
+
+**Step 10:** Open `PCbuild/python.props` and change the entries as shown below:
+
+![](/img/openssl_settings2.jpg)
+
+**Step 11:** Open the Python Solution under `PCbuild/pcbuild.sln` in VS 2015.
+Change the Linker settings of `_hashlib` and `_ssl` projects as shown below:
+
+![](/img/openssl_vs_settings.jpg)
+
+**Step 12:** Now, build `_hashlib.pyd` and `_ssl.pyd` in VS 2015.  
+**Step 13:** Copy these built .pyd files from `\PCbuild\amd64\` to a Python binary installation directory `C:\python311\DLLs`.  
+**Step 14:** Start Python and use these commands to check the OpenSSL 3.1 version.  
+
+![](/img/openssl-after.jpg)
+
+**Step 15:** Now, to run Python with a FIPS enabled OpenSSL, create `openssl.cnf` and `fipsmodule.cnf` file using the below content in `C:\Python311` directory. By using this config file, FIPS will be enabled by default.
+
+> #### openssl.cnf:
```

Comment:
Please try to use a code block snippet for the openssl.cnf file. Thsi would allow the readers to copy/paste the content.